app/aws/s3.py

```diff
@@ -132,33 +133,74 @@ def cleanup_old_s3_objects():
+    and the personalization through job_cache[<job_id>_personalisation], which
+    in theory should make report generation a lot faster.
+
+    We are moving processing from the front end where the user can see it
+    in wait time, to this back end process.
+    """
+    try:
+        job_id = get_job_id_from_s3_object_key(object_key)
+        if job_cache.get(job_id) is None:
+            object = (
+                s3res.Object(bucket_name, object_key)
+                .get()["Body"]
+                .read()
+                .decode("utf-8")
+            )
+            # TODO we probably don't need this phone number check anymore.
+            # Originally, there were csv uploads that were not valid
+            # for messaging sending.  They may have been experimental, or
+            # they may have not been caught by validation.  When we're sure
+            # all csv files have a phone number column, we can remove the if.
+            # We are essentially just making sure this object looks like a job.
+            if "phone number" in object.lower():
+                set_job_cache(job_cache, job_id, object)
+                set_job_cache(job_cache, f"{job_id}_phones", extract_phones(object))
+                set_job_cache(
+                    job_cache,
+                    f"{job_id}_personalisation",
+                    extract_personalisation(object),
+                )
+
+    except LookupError:
+        # perhaps our key is not formatted as we expected.  If so skip it.
+        current_app.logger.exception("LookupError #notify-admin-1200")
 
+
+def get_s3_files():
+    """
+    We're using the ThreadPoolExecutor here to speed up the retrieval of S3
+    csv files for scaling needs.
+    """
     bucket_name = current_app.config["CSV_UPLOAD_BUCKET"]["bucket"]
-    objects = list_s3_objects()
+    object_keys = list_s3_objects()
 
     s3res = get_s3_resource()
     current_app.logger.info(
         f"job_cache length before regen: {len(job_cache)} #notify-admin-1200"
     )
-    for object in objects:
-        # We put our csv files in the format "service-{service_id}-notify/{job_id}"
-        try:
-            object_arr = object.split("/")
-            job_id = object_arr[1]  # get the job_id
-            job_id = job_id.replace(".csv", "")  # we just want the job_id
-            if job_cache.get(job_id) is None:
-                object = (
-                    s3res.Object(bucket_name, object)
-                    .get()["Body"]
-                    .read()
-                    .decode("utf-8")
-                )
-                if "phone number" in object.lower():
-                    set_job_cache(job_cache, job_id, object)
-        except LookupError:
-            # perhaps our key is not formatted as we expected.  If so skip it.
-            current_app.logger.exception("LookupError #notify-admin-1200")
+    try:
+        with ThreadPoolExecutor() as executor:
```

Comment:
Is there anything we ought to account for here with the `max_workers` argument?  [According to the docs](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor):

>Default value of max_workers is changed to min(32, os.cpu_count() + 4). This default value preserves at least 5 workers for I/O bound tasks. It utilizes at most 32 CPU cores for CPU bound tasks which release the GIL. And it avoids using very large resources implicitly on many-core machines.
>
>ThreadPoolExecutor now reuses idle worker threads before starting max_workers worker threads too.

Perhaps it's safe enough as is according to those notes, but I just want to make sure we're thinking through things fully and don't have runaway resource usage!

Reply:
I think we are okay but time will tell